test/unit/BPool.t.sol

```diff
@@ -1433,31 +1433,173 @@ contract BPool_Unit_JoinswapPoolAmountOut is BasePoolTest {
+    _fuzz.poolAmountOut = 0;
 
-  function test_Revert_Reentrancy() private view {}
+    vm.expectRevert('ERR_MATH_APPROX');
+    bPool.joinswapPoolAmountOut(tokenIn, _fuzz.poolAmountOut, type(uint256).max);
+  }
 
-  function test_Set_Balance() private view {}
+  function test_Revert_LimitIn(JoinswapPoolAmountOut_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    uint256 _tokenAmountIn = calcSingleInGivenPoolOut(
+      _fuzz.tokenInBalance,
+      _fuzz.tokenInDenorm,
+      _fuzz.totalSupply,
+      _fuzz.totalWeight,
+      _fuzz.poolAmountOut,
+      _fuzz.swapFee
+    );
 
-  function test_Emit_LogJoin() private view {}
+    vm.expectRevert('ERR_LIMIT_IN');
+    bPool.joinswapPoolAmountOut(tokenIn, _fuzz.poolAmountOut, _tokenAmountIn - 1);
+  }
 
-  function test_Mint_PoolShare() private view {}
+  function test_Revert_MaxInRatio(JoinswapPoolAmountOut_FuzzScenario memory _fuzz) public {
+    // Replicating _assumeHappyPath, but removing irrelevant assumptions and conditioning the revert
+    _fuzz.tokenInDenorm = bound(_fuzz.tokenInDenorm, MIN_WEIGHT, MAX_WEIGHT);
+    _fuzz.swapFee = bound(_fuzz.swapFee, MIN_FEE, MAX_FEE);
+    _fuzz.totalWeight = bound(_fuzz.totalWeight, MIN_WEIGHT * MAX_BOUND_TOKENS, MAX_WEIGHT * MAX_BOUND_TOKENS);
+    _fuzz.tokenInBalance = bound(_fuzz.tokenInBalance, MIN_BALANCE, type(uint256).max / MAX_IN_RATIO);
+    vm.assume(_fuzz.totalSupply >= INIT_POOL_SUPPLY);
+    vm.assume(_fuzz.totalSupply < type(uint256).max - _fuzz.poolAmountOut);
+    _assumeCalcSingleInGivenPoolOut(
+      _fuzz.tokenInBalance,
+      _fuzz.tokenInDenorm,
+      _fuzz.totalSupply,
+      _fuzz.totalWeight,
+      _fuzz.poolAmountOut,
+      _fuzz.swapFee
+    );
+    uint256 _tokenAmountIn = calcSingleInGivenPoolOut(
+      _fuzz.tokenInBalance,
+      _fuzz.tokenInDenorm,
+      _fuzz.totalSupply,
+      _fuzz.totalWeight,
+      _fuzz.poolAmountOut,
+      _fuzz.swapFee
+    );
+    vm.assume(_tokenAmountIn > bmul(_fuzz.tokenInBalance, MAX_IN_RATIO));
 
-  function test_Push_PoolShare() private view {}
+    _setValues(_fuzz);
 
-  function test_Pull_Underlying() private view {}
+    vm.expectRevert('ERR_MAX_IN_RATIO');
+    bPool.joinswapPoolAmountOut(tokenIn, _fuzz.poolAmountOut, type(uint256).max);
+  }
 
-  function test_Returns_TokenAmountIn() private view {}
+  function test_Revert_Reentrancy() public {
+    // Assert that the contract is accesible
+    assertEq(bPool.call__mutex(), false);
 
-  function test_Emit_LogCall() private view {}
+    // Simulate ongoing call to the contract
+    bPool.set__mutex(true);
+
+    vm.expectRevert('ERR_REENTRY');
+    bPool.joinswapPoolAmountOut(tokenIn, 0, 0);
+  }
+
+  function test_Set_Balance(JoinswapPoolAmountOut_FuzzScenario memory _fuzz) public happyPath(_fuzz) {
+    uint256 _balanceBefore = bPool.getBalance(tokenIn);
+
+    bPool.joinswapPoolAmountOut(tokenIn, _fuzz.poolAmountOut, type(uint256).max);
```

Comment:
Here I propose the same as in https://github.com/defi-wonderland/balancer-v1-amm/pull/32#discussion_r1616558393 in relation with the input arguments.